### ISSUE TITLE
Question about PXN all2all

### ISSUE THREAD

**Issue by ConnollyLeon** [NVIDIA/nccl#1030]
Hi NCCL team. I have read your blog [doubling all2all performance with nccl 2.12](https://developer.nvidia.com/blog/doubling-all2all-performance-with-nvidia-collective-communication-library-2-12). 

I wonder if NCCl would do both message aggregation and PXN automatically using this [all2all implementation](https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/usage/p2p.html#all-to-all)?

I guess it is kinda like an intra-node all2all first and then inter-node all2all.

Looking forward to your replies. ;)

**Comment by sjeaugey** [comment]
The alltoall example above is just an alltoall, including all peers, intra- and inter-node. That would use PXN when possible (on NVLink platforms).

**Comment by shanleo2024** [comment]
> The alltoall example above is just an alltoall, including all peers, intra- and inter-node. That would use PXN when possible (on NVLink platforms).

Hi @sjeaugey I cannot understand the PXN Message aggregation, the bolg says PXN will use a new multireceive function, what is the new multireceive function?
`With PXN, all GPUs on a given node move their data onto a single GPU for a given destination. This enables the network layer to aggregate messages, by implementing a new multireceive function. `
Is the new multireceive function from net plugs-in?
Which NCCL version support this new multireceive function?
Thank you.

**Comment by sjeaugey** [comment]
Yes the multi-receive function is implemented in the network plugin. It was implemented in NCCL 2.12 so it's pretty old now.

**Comment by shanleo2024** [comment]
> Yes the multi-receive function is implemented in the network plugin. It was implemented in NCCL 2.12 so it's pretty old now.

@sjeaugey Are you saying that this feature is only implemented in the net plugin? Has it also been implemented in NCCL IB verbs?
I guess in the function recvProxyProgress, the subCount represents the multi-receive, but not make sure, please correct me .

`NCCLCHECK(proxyState->ncclNet->irecv(resources->netRecvComm, subCount, ptrs, sizes, tags, mhandles, requestPtr));`

Thank you a lot!

**Comment by sjeaugey** [comment]
> @sjeaugey Are you saying that this feature is only implemented in the net plugin? Has it also been implemented in NCCL IB verbs?

When I say "plugin" I also include internal plugins, i.e. the IB verbs plugin or the socket plugin (although the socket plugin doesn't implement PXN).

> I guess in the function recvProxyProgress, the subCount represents the multi-receive

Yes indeed, subCount is the number of receives.

**Comment by shanleo2024** [comment]
@sjeaugey According https://developer.nvidia.com/blog/doubling-all2all-performance-with-nvidia-collective-communication-library-2-12/#entry-content-comments
```
Message aggregation

With PXN, all GPUs on a given node move their data onto a single GPU for a given destination. This enables the network layer to aggregate messages, by implementing a new multireceive function. The function enables the remote CPU proxy to send all messages as one as soon as they are all ready.

For example, if a GPU on a node is performing an all2all operation and is to receive data from all eight GPUs from a remote node, NCCL calls a multireceive with eight buffers and sizes. On the sender side, the network layer can then wait until all eight sends are ready, then send all eight messages at one time, which can have a significant effect on the message rate.
```
So the multireceive function is  implemented by proxyState->ncclNet->irecv, right.

But which confused me is that:
`The function enables the remote CPU proxy to send all messages as one as soon as they are all ready`

I noticed the remote CPU proxy will send messages one by one  in:
`proxyState->ncclNet->isend(resources->netSendComm, buff, size, resources->tpRank, sub->mhandle, sub->requests+buffSlot)`
As here buff and size represents one message.

How to understand `the remote CPU proxy to send all messages as one as soon as they are all ready`?

Thank you again.

**Comment by shanleo2024** [comment]
Hi @sjeaugey,
From my understanding, the remote CPU proxy to send all messages as one as soon as they are all ready is implemented by ncclIbMultiSend, all of the messages will be packeted in a list of wrs, and only call the ibv_post_send once.
Is my understanding right?
Thank you.

**Comment by sjeaugey** [comment]
That is correct indeed. On the receiver side, the NCCL core can indicateto the network plugin that it should receive 8 messages at once, but on the sender side each message is coming from a different rank hence it is part of a different proxy operation (can't group them together), so the aggregation happens inside the network plugin.
